ath79: convert some ath10k calibration data to NVMEM format

DragonBluep opened the pull request:
This PR will convert ath10k calibration data to NVMEM format for most ath79 target devices.

Ansuel commented:
Oh no... Another "Paper Please" Session :'(

![image](https://github.com/openwrt/openwrt/assets/20289090/f8d0a7da-279a-4b68-b5a3-cbe975a6d171)

Ansuel commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -3,200 +3,12 @@
-	yuncore,a782|\
-	yuncore,xd4200)
-		caldata_extract "art" 0x5000 0x2f20
-		ln -sf /lib/firmware/ath10k/pre-cal-pci-0000\:00\:00.0.bin \
```
Any idea what is the usage of this link? did we use the same calibration for precal and cal?

DragonBluep replied:
This link seems a copy from ath10k-ct-firmware package. They are useless here.
https://github.com/openwrt/openwrt/blob/920414ca8848fe1b430e436207b4f8c927819368/package/firmware/ath10k-ct-firmware/Makefile#L594-L620

I think those ct-firmware packages need some fix because it will create non-exist soft link if device use nvmem calibration data. And `board-2.bin` seems also useless for ath79 target. I remember only ipq401x devices require it. If I'm wrong please correct me.
```
root@OpenWrt:~# ls -al /lib/firmware/ath10k/QCA9888/hw2.0/
drwxr-xr-x    2 root     root            73 Jan 28 10:10 .
drwxr-xr-x    3 root     root            28 Jan 28 10:10 ..
-rw-r--r--    1 root     root        206444 Jan 28 10:10 board-2.bin
lrwxrwxrwx    1 root     root            30 Jan 28 10:10 board.bin -> ../../cal-pci-0000:01:00.0.bin
-rw-r--r--    1 root     root        630784 Jan 28 10:10 firmware-5.bin
root@OpenWrt:~# cd /lib/firmware/ath10k/QCA9888/hw2.0/
root@OpenWrt:/lib/firmware/ath10k/QCA9888/hw2.0# cat '../../cal-pci-0000:01:00.0.bin'
cat: can't open '../../cal-pci-0000:01:00.0.bin': No such file or directory
```

DragonBluep replied:
> did we use the same calibration for precal and cal?

In this script, at least they did the same thing. Read cal/pre-cal data from MTD and then save .bin file to the `lib/firmware/ath10k` folder. The `ln` command for ath10k wave2 devices did not really take effect.

Ansuel replied:
Ok ath10k fw load is complex enough to make some confusion but from what I notice in the code:
- pre-cal IS OPTIONAL, is loaded from nvmem or tried to be fetched from firmware directory following `pre-cal-%s-%s.bin` 
- cal IS MANDATORY and probe will fail without it. We have board-2.bin

It seems the link was used just to handle the cal file being mandatory for device that didn't had board-2 by using the pre-cal as board file. (on these device pre-cal is API 1)

I guess that is not the case anymore and we always ship board-2.bin on any device that use ath10k correct?

Ansuel commented on target/linux/ath79/dts/qca9531_yuncore_a770.dts:
```diff
@@ -123,6 +127,13 @@
+		compatible = "qcom,ath10k";
+		reg = <0x0000 0 0 0 0>;
+		nvmem-cells = <&cal_art_5000>;
+		nvmem-cell-names = "calibration";
```
here we used calibration but in the other case we use pre-calibration? Think we have to use calibration in these strange case to remove any chance of causing regression...

Ansuel commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -3,200 +3,12 @@
-		caldata_extract "art" 0x5000 0x844
-		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary info 0x8) 2)
-		;;
-	tplink,re355-v1|\
-	tplink,re450-v1)
-		caldata_extract "art" 0x5000 0x844
-		ath10k_patch_mac $(macaddr_add $(cat /sys/class/net/eth0/address) -2)
-		;;
```
think we are missing these? and we are setting the addr from the info partition (from rxc5x dtsi?)

DragonBluep replied:
The `eth0` MAC is fetched from info mtd. So, we can configure it directly in dts.
https://github.com/openwrt/openwrt/blob/3a073a021286bef3fed0ca7884872c5890420d60/target/linux/ath79/dts/qca9558_tplink_rex5x.dtsi#L173

Ansuel commented on target/linux/ath79/dts/qca9561_nec_wf1200cr.dts:
```diff
@@ -156,6 +166,8 @@
 		compatible = "qcom,ath10k";
 		reg = <0x0000 0 0 0 0>;
+		nvmem-cells = <&precal_art_5000>;
+		nvmem-cell-names = "pre-calibration";
```
Same concern pre-cal instead of cal?

Ansuel commented:
Ok I checked all the dts change and we just need to understand the pre-cal situation and how to handle it.

DragonBluep commented:
> Ok I checked all the dts change and we just need to understand the pre-cal situation and how to handle it.

Thanks for your review. I have both wave1(cal) and wave2(pre-cal) boards so I can do some tests to check it, maybe on this weekend. But before this, I found this commit https://github.com/openwrt/openwrt/commit/7154a1dc7b9f1b01f30ef0281e65c952c89d375b. So I guess wave2 boards must use `pre-calibration`. @svanheule should know more about this.

svanheule commented:
I don't really know why wave2 devices need the pre-cal data, it was more of "it doesn't work without it". These ath79 devices were of the first that I ported, so I mostly just did what devices with the same radios were doing... Sorry I can't be of any more help!

Ansuel commented:
@DragonBluep I need some info on what is the actual problem with them. Probe error? Wifi doesn't work? Anyway as long as we embed board-2.bin in those fw then we can use pre-cal... Again we just need to know why the symbolic link was created for these devices...

DragonBluep commented:
> @DragonBluep I need some info on what is the actual problem with them. Probe error? Wifi doesn't work? Anyway as long as we embed board-2.bin in those fw then we can use pre-cal... Again we just need to know why the symbolic link was created for these devices...

Hi! For this question, I think I can come to a simple conclusion now. **The `board.bin` symbolic link is useless if `board-2.bin` exists**, as you said before.

Test1. fetch data via ath10k firmware script. Everything works.
```
root@OpenWrt:~# ls /lib/firmware/ath10k/
QCA9888/                      pre-cal-pci-0000:00:00.0.bin
root@OpenWrt:~# ls /lib/firmware/ath10k/QCA9888/hw2.0/
board-2.bin     board.bin       firmware-5.bin
root@OpenWrt:~# dmesg |grep ath10k
[    9.476301] ath10k 6.4 driver, optimized for CT firmware, probing pci device: 0x56.
[    9.510915] ath10k_pci 0000:00:00.0: enabling device (0000 -> 0002)
[    9.517689] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   13.737214] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
[   13.746804] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   13.767232] ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
[   14.136591] ath10k_pci 0000:00:00.0: board_file api 2 bmi_id 0:17 crc32 ae4cbea5
[   15.900710] ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   15.908514] ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
[   15.938909] ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   15.950228] ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
[   16.185568] ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[   68.823191] ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   68.831045] ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
[   68.861415] ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   68.872764] ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
[   69.185818] ath10k_pci 0000:00:00.0: rts threshold -1
[   69.195417] ath10k_pci 0000:00:00.0: Firmware lacks feature flag indicating a retry limit of > 2 is OK, requested limit: 4
[   69.822336] ath10k_pci 0000:00:00.0: NOTE:  Firmware DBGLOG output disabled in debug_mask: 0x10000000
[  941.617332] ath10k_pci 0000:00:00.0: mac flush vdev 0 drop 0 queues 0x1 ar->paused: 0x0  arvif->paused: 0x0
```

Test2. fetch data via ath10k firmware script, without `board.bin` link. Everything works.
```
root@OpenWrt:~# ls /lib/firmware/ath10k/QCA9888/hw2.0
board-2.bin     firmware-5.bin
root@OpenWrt:~# dmesg |grep ath10
[    9.636134] ath10k 6.4 driver, optimized for CT firmware, probing pci device: 0x56.
[    9.670886] ath10k_pci 0000:00:00.0: enabling device (0000 -> 0002)
[    9.677663] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   13.666991] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
[   13.676585] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   13.697001] ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
[   14.066406] ath10k_pci 0000:00:00.0: board_file api 2 bmi_id 0:17 crc32 ae4cbea5
[   15.830554] ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   15.838364] ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
[   15.868728] ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   15.880064] ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
[   16.114772] ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[   45.123207] ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   45.131055] ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
[   45.161420] ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   45.172760] ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
[   45.485459] ath10k_pci 0000:00:00.0: rts threshold -1
[   45.491252] ath10k_pci 0000:00:00.0: Firmware lacks feature flag indicating a retry limit of > 2 is OK, requested limit: 4
[   46.122395] ath10k_pci 0000:00:00.0: NOTE:  Firmware DBGLOG output disabled in debug_mask: 0x10000000
```

The driver will search `board-2.bin` first, if file does not exist, then it will try `board.bin`.
https://github.com/greearb/ath10k-ct/blob/fadd0768cbd22248a60efbb219ccefc9d86cd78c/ath10k-6.4/core.c#L1991-L2001


As for why Comfast series devices removed the `board-2.bin`, this is the original PR https://github.com/openwrt/openwrt/pull/1942. The author said `I had to add "rm /lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin", since
the driver tries to load board-2.bin (the default calibration data from the firmware package) first and, only when that file is not found, falls back to board.bin (the calibration data read from the flash).` But My opinion is that it is not true. The driver will load both `board.bin` (aka pci-xxx.bin in firmware/ath10k directory) and `board-2.bin`. The evidence is that if I don't provide the correct pre-calibration data (board.bin/pre-calibration nvmem), the driver won't run at all.

Test3. use `calibration` nvmem cell to provide wrong calibration data for wave 2 device. Fail.
```
[    9.523543] ath10k 6.4 driver, optimized for CT firmware, probing pci device: 0x56.
[    9.560964] ath10k_pci 0000:00:00.0: enabling device (0000 -> 0002)
[    9.567749] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   14.314859] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
[   14.324445] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   14.344876] ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
[   14.684263] ath10k_pci 0000:00:00.0: failed to fetch board data for bus=pci,vendor=168c,device=0056,subsystem-vendor=0000,subsystem-device=0000 from ath10k/QCA9888/hw2.0/board-2.bin
[   15.348584] ath10k_pci 0000:00:00.0: failed to fetch board-2.bin or board.bin from ath10k/QCA9888/hw2.0
[   15.358360] ath10k_pci 0000:00:00.0: failed to fetch board file: -12
[   15.380091] ath10k_pci 0000:00:00.0: could not probe fw (-12)
```

The final conclusions:
1. This PR won't introduce new strange issues.
2. The symbolic link is essential for Comfast routers because they don't have `board-2.bin`. But the right fix is adding back `board-2.bin` instead of creating a `board.bin` link, just like other devices. I can add another commit to do that first before nvmem converting.
3. The symbolic link code can be removed from ath10k-ct firmware package and ath79/ipq40xx targets because we always install `board-2.bin` now. Perhaps we don't ship `board-2.bin` years ago? Not sure.

Ansuel commented:
> 2\. The symbolic link is essential for Comfast routers because they don't have `board-2.bin`. But the right fix is adding back `board-2.bin` instead of creating a `board.bin` link, just like other devices. I can add another commit to do that first before nvmem converting.

we can consider defining calibration without adding pre-cal to it. Should result in the same thing in theory. If this work then we might apply this logic to everything that doesn't have board-2. 

The link is needed if for some reason board-2 doesn't have the board id in it but it seems unlikely.

DragonBluep commented:
> > 2. The symbolic link is essential for Comfast routers because they don't have `board-2.bin`. But the right fix is adding back `board-2.bin` instead of creating a `board.bin` link, just like other devices. I can add another commit to do that first before nvmem converting.
> 
> we can consider defining calibration without adding pre-cal to it. Should result in the same thing in theory. If this work then we might apply this logic to everything that doesn't have board-2.
> 

If there are any patches that I can test, please let me know.

> The link is needed if for some reason board-2 doesn't have the board id in it but it seems unlikely.

BTW, in this case, we can't use nvmem because nvmem won't create firmware/ath10k/precal-xxx.bin so we can't use `board.bin`.

Ansuel commented:
> If there are any patches that I can test, please let me know.

no patch just change from pre-calibration to calibration for the nvmem and see if it does work.

DragonBluep commented:
> > If there are any patches that I can test, please let me know.
> 
> no patch just change from pre-calibration to calibration for the nvmem and see if it does work.

No. The driver won't up. Please see Test3.
> Test3. use calibration nvmem cell to provide wrong calibration data for wave 2 device. Fail. 

The same result as svanheule said.
> I don't really know why wave2 devices need the pre-cal data, it was more of "it doesn't work without it".

Ansuel commented:
Well I guess we have to provide board-2 and hope for the best...

DragonBluep commented:
Update: Add board-2.bin to COMFAST devices.

DragonBluep commented:
> @DragonBluep also I know it might be annoying for you but I wonder if you can split this in various subcommit... one for the simple devices that only took the calibration... one for the one where mac and precal has to be fixed... (it would be easier to handle reverting the single commit if we notice something wrong in it)

Thank you for your suggestion. I will do it later this week.

DragonBluep committed:
ath79: convert ath10k calibration data to NVMEM (ASCII MAC)

This patch converts ath10k calibration data to NVMEM format for
wave 1 devices with mtd ASCII MAC address. The "calibration"
NVMEM cell size is 0x844. All unportable MAC address settings
have been moved to '10_fix_wifi_mac' scripts.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ath79: convert ath10k calibration data to NVMEM (binary MAC)

This patch converts ath10k calibration data to NVMEM format for
wave 1 devices with mtd binary MAC address. The "calibration"
NVMEM cell size is 0x844. The MAC addresses are assigned via dts.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ath79: convert ath10k pre-calibration data to NVMEM (binary MAC)

This patch converts ath10k pre-calibration data to NVMEM format for
wave 2 devices with mtd binary MAC address. The "pre-calibration"
NVMEM cell size is 0x2f20. The MAC addresses are assigned via dts.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ath79: add back board-2.bin to COMFAST devices

The ath10k driver will load both pre-calibration data and board-2.bin
if board-2.bin exists. So it's not necessary to remove it. And this
change won't increase jffs2 image size.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ath79: convert ath10k pre-calibration data to NVMEM (built-in MAC)

This patch converts ath10k pre-calibration data to NVMEM format for
wave 2 devices with built-in MAC address. The "pre-calibration"
NVMEM cell size is 0x2f20.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ath79: convert ath10k pre-calibration data to NVMEM (ASCII MAC)

This patch converts ath10k pre-calibration data to NVMEM format for
wave 2 devices with mtd ASCII MAC address. The "pre-calibration"
NVMEM cell size is 0x2f20. All unportable MAC address settings have
been moved to '10_fix_wifi_mac' scripts.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ath79: convert ath10k calibration data to NVMEM (built-in MAC)

This patch converts ath10k calibration data to NVMEM format for
wave 1 devices with built-in MAC address. The "calibration"
NVMEM cell size is 0x844.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
Update:
Each conversion commit is divided into three parts, built-in MAC devices, binary MAC devices and ASCII MAC devices.

testuser7 commented:
For `ath9k`, should the `mtd-cal-data` call also be replaced with NVMEM format?

DragonBluep commented:
> For `ath9k`, should the `mtd-cal-data` call also be replaced with NVMEM format?

Yep. @neheb has already started this work in https://github.com/openwrt/openwrt/pull/14082.

castillofrancodamian commented:
The 5 GHz radio is not detected. [More info here](https://github.com/openwrt/openwrt/issues/14541).
